build.xml

```diff
@@ -121,6 +123,14 @@
   <target name="-real-build">
     <mkdir dir="${build-src.dir}"/>
 
+    <!-- Make sure that the bootclasspath isn't newer than expected. -->
+    <available property="compile.java.isNewer"
+      classname="${compile.java.newerClassname}"
+      classpath="${compile.java.bootclasspath}" ignoresystemclasses="true"/>
+    <fail if="compile.java.isNewer">
+      Error: build.bootclasspath is newer than JDK ${compile.java.version}.
```

Comment:
I don't know what this situation means. If the source code compiles with JDK 6 it will also compile with JDK 8.  Are we trying to prevent people from using a new JDK?

Reply:
This is the thing where the target and bootclasspath must be set together. Setting them separately is error prone. There are several situations handled by this check:

1. If you compile with a later JDK, and build.bootclasspath is not set, then javac will issue a warning to that effect, since the target is set. You'll get a JAR file that may or may not work on the targeted JRE, but which will always work on the build JDK.

2. If you compile with a later JDK, and set build.bootclasspath to point to Java classes of the correct version, then you'll get a JAR file that is guaranteed to work on the target JVM. This is what we want.

3. If you compile with a later JDK, and set build.bootclasspath to point to Java classes also from a later JDK, this check will stop the build, rather than produce a JAR file that may not work on the target JVM. javac doesn't detect this case, but it's the same problem as case 1. I made it an error because we really don't want to build like this.

Reply:
OK.  I think I follow.  Question to double check: Why not treat case #3 like case #1.  In both cases the bootclasspath is not the target JRE bootclasspath.

Another follow up, in order to do things safely, someone needs to have say both JDK8 and JDK6 (or at least JDK8 and JRE6)?

Final followup, the way I do things while developing, I'm in danger case #1. The way I realease things by saying JAVA_HOME=$JDK6 I am in safe case for release. Yes?

Reply:
+1 for adding the entire github review comment you wrote above with the three cases into the build.xml file.

Reply:
I made case 3 an error instead of a warning because the person explicitly tried to do the correct thing but failed. I can make it a warning if you prefer.

You just need the classes themselves, and not the whole JRE. On my MacBooks over the years, upgrading Java would sometimes uninstall the old version(s), so I keep the JAR files from previous releases around. I currently only have JDK 8 installed, for example.

In Java 9, all this machinery (-source, -target, -bootclasspath, and needing the older classes) has been replaced by -release. We could, instead, require production builds to use Java 9, and use -release, and if you're not using Java 9, you get whatever random thing your environment produces. Today, we require production builds to set build.bootclasspath to the right JAR files, and if you don't, you get whatever random thing you get.

Yes, you are safe. However, if you set build.bootclasspath in the first place, then it doesn't ever matter what you compile with, invalid code can't sneak into the code base, and you have the option of avoiding compiler bugs.

How would you feel about putting the expanded explanations into the commit message instead?